src/vset.c

```diff
@@ -0,0 +1,1249 @@
+            if (new_bucket != bucket)
+                /* In order to avoid rax override, we directly change the node data */
+                // alternative: raxInsert(expiry_buckets, key, key_len, new_bucket, NULL);
+                raxSetData(node, new_bucket);
+        }
+    } else if (vsetBucketType(bucket) == VSET_BUCKET_HT) {
+        bucket = insertToBucket_HASHTABLE(getExpiry, bucket, entry, expiry);
+    } else {
+        panic("Unknown bucket type in insertToBucket_RAX");
+    }
+    return target;
+}
+
+static inline vsetBucket *removeFromBucket_SINGLE(vsetGetExpiryFunc getExpiry, vsetBucket *bucket, void *entry, long long expiry, bool *removed) {
+    UNUSED(getExpiry);
+    UNUSED(expiry);
+
+    if (vsetBucketSingle(bucket) == entry) {
+        *removed = true;
+        return vsetBucketSetNone(bucket);
+    } else {
+        *removed = false;
+        return bucket;
+    }
+}
+
+static inline vsetBucket *removeFromBucket_VECTOR(vsetGetExpiryFunc getExpiry, vsetBucket *bucket, void *entry, long long expiry, bool *removed) {
+    UNUSED(getExpiry);
+    UNUSED(expiry);
+
+    vsetBucket *new_bucket = bucket;
+    bool success = false;
+    pVector *sv = vsetBucketVector(bucket);
+    /* In case we we removed the entry */
+    uint32_t vlen = pvLen(sv);
+    if (vlen <= 2) {
+        /* convert to single if needed */
+        uint32_t idx = pvFind(sv, entry);
+        if (idx == vlen) {
+            success = false;
+        } else {
+            if (vlen == 1)
+                new_bucket = vsetBucketSetNone(bucket);
+            else
+                new_bucket = vsetBucketSetSingle(bucket, pvGet(sv, idx == 0 ? 1 : 0));
+            success = true;
+            pvFree(sv);
+        }
+    } else {
+        if (pvRemove(&sv, entry)) {
+            success = true;
+            new_bucket = vsetBucketSetVector(bucket, sv);
+        }
+    }
+    if (removed) *removed = success;
+    return new_bucket;
+}
+
+static inline vsetBucket *removeFromBucket_HASHTABLE(vsetGetExpiryFunc getExpiry, vsetBucket *bucket, void *entry, long long expiry, bool *removed) {
+    UNUSED(getExpiry);
+    UNUSED(expiry);
+
+    bool success = false;
+    vsetBucket *new_bucket = bucket;
+    hashtable *ht = vsetBucketHashtable(bucket);
+    if (hashtableDelete(ht, entry)) {
+        success = true;
+        assert(hashtableSize(ht) > 0);
+        if (hashtableSize(ht) == 1) {
+            // Downgrade to SINGLE
+            hashtableIterator hi;
+            hashtableInitIterator(&hi, ht, 0);
+            void *ptr;
+            hashtableNext(&hi, &ptr);
+            hashtableRelease(ht);
+            new_bucket = vsetBucketSetSingle(bucket, ptr);
+        }
+    }
+    if (removed) *removed = success;
+    return new_bucket;
+}
+static bool raxBucketRemoveEntry(vsetBucket *parent, vsetGetExpiryFunc getExpiry, void *entry, vsetBucket *bucket, unsigned char *key, size_t key_len, vsetBucket **pbucket, raxNode *node) {
```

Comment:
Minor naming: The rax prefix hints that this function part of the API defined in rax.{c,h} which it isn't. I would avoid the rax prefix for this reason. Maybe just start with `remove` as in the other similar internal functions?

Not sure if the suggestion below makes sense. Is it child a bucket? They can't be nested but there can be two levels only? I see the risk of confusion with removeFromBucket_RAX so I can see it shouldn't be named too similar to that. Btw, we use rax for two different things which I think needs to be explained in the top comment and/or vset introduction code comments. There is some risk of confusion between the two.

```suggestion
static bool removeChildBucketFromRax(vsetBucket *rax_bucket, vsetGetExpiryFunc getExpiry, void *entry, vsetBucket *bucket, unsigned char *key, size_t key_len, vsetBucket **pbucket, raxNode *node) {
```